src/query/api/v1/handler/prometheus/native/read.go

```diff
@@ -118,129 +109,130 @@ type RespError struct {
+// ParseRequest parses the given request.
+func ParseRequest(
+	ctx context.Context,
+	r *http.Request,
+	opts options.HandlerOptions,
+) (ParsedOptions, *xhttp.ParseError) {
+	fetchOpts, rErr := opts.FetchOptionsBuilder().NewFetchOptions(r)
 	if rErr != nil {
-		h.promReadMetrics.fetchErrorsClient.Inc(1)
-		return nil, emptyReqParams, &RespError{Err: rErr.Inner(), Code: rErr.Code()}
+		return ParsedOptions{}, rErr
 	}
 
-	if params.Debug {
-		logger.Info("request params", zap.Any("params", params))
+	queryOpts := &executor.QueryOptions{
+		QueryContextOptions: models.QueryContextOptions{
+			LimitMaxTimeseries: fetchOpts.Limit,
+		}}
+
+	restrictOpts := fetchOpts.RestrictQueryOptions.GetRestrictByType()
+	if restrictOpts != nil {
+		restrict := &models.RestrictFetchTypeQueryContextOptions{
+			MetricsType:   uint(restrictOpts.MetricsType),
+			StoragePolicy: restrictOpts.StoragePolicy,
+		}
+
+		queryOpts.QueryContextOptions.RestrictFetchType = restrict
 	}
 
-	if err := h.validateRequest(&params); err != nil {
-		h.promReadMetrics.fetchErrorsClient.Inc(1)
-		return nil, emptyReqParams, &RespError{Err: err, Code: http.StatusBadRequest}
+	engine := opts.Engine()
+	params, rErr := parseParams(r, engine.Options(),
+		opts.TimeoutOpts(), fetchOpts, opts.InstrumentOpts())
+	if rErr != nil {
+		return ParsedOptions{}, rErr
 	}
 
-	result, err := read(ctx, engine, queryOpts, fetchOpts, h.tagOpts,
-		w, params, h.instrumentOpts)
-	if err != nil {
-		sp := xopentracing.SpanFromContextOrNoop(ctx)
-		sp.LogFields(opentracinglog.Error(err))
-		opentracingext.Error.Set(sp, true)
-		logger.Error("range query error",
-			zap.Error(err),
-			zap.Any("params", params),
-			zap.Any("queryOpts", queryOpts),
-			zap.Any("fetchOpts", fetchOpts))
-		h.promReadMetrics.fetchErrorsServer.Inc(1)
-		return nil, emptyReqParams, &RespError{
-			Err:  err,
-			Code: http.StatusInternalServerError,
-		}
+	maxPoints := opts.Config().Limits.MaxComputedDatapoints()
+	if err := validateRequest(&params, maxPoints); err != nil {
+		return ParsedOptions{}, xhttp.NewParseError(err, http.StatusBadRequest)
 	}
 
-	// TODO: Support multiple result types
-	w.Header().Set("Content-Type", "application/json")
-	handleroptions.AddWarningHeaders(w, result.meta)
-	return result.series, params, nil
+	return ParsedOptions{
+		QueryOpts: queryOpts,
+		FetchOpts: fetchOpts,
+		Params:    params,
+	}, nil
 }
 
-func (h *PromReadHandler) validateRequest(params *models.RequestParams) error {
-	// Impose a rough limit on the number of returned time series. This is intended to prevent things like
-	// querying from the beginning of time with a 1s step size.
-	// Approach taken directly from prom.
-	numSteps := int64(params.End.Sub(params.Start) / params.Step)
-	maxComputedDatapoints := h.limitsCfg.MaxComputedDatapoints()
-	if maxComputedDatapoints > 0 && numSteps > maxComputedDatapoints {
+func validateRequest(params *models.RequestParams, maxPoints int) error {
```

Comment:
Does `params` necessarily need to be a pointer type?  Could it be just `params models.RequestParams` (I notice we do `&params` when we call this which might be unnecessary)?

Reply:
Don't think so; just kept the types as is generally